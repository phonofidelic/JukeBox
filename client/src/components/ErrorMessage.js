import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogTitle';
import DialogActions from '@material-ui/core/DialogActions';
import Typography from '@material-ui/core/Typography';

class ErrorMessage extends Component {
	state = {
		errorDismissed: false
	}

	render() {
		const { 
			error, 
			handleClearError,
		} = this.props;

		return (
			<div>
				<Dialog open={error.showError}>
					<DialogTitle>{error.title}</DialogTitle>
					<DialogContent>
						<Typography>{error.status && `${error.status} - `}{error.message}</Typography>
					</DialogContent>
					<DialogActions>
						<Button 
							data-cy="error-dismiss-button"
							variant="outlined" 
							onClick={() => handleClearError()}
						>
							Ok
						</Button>
					</DialogActions>
				</Dialog>
			</div>
		);
	}
}

ErrorMessage.propTypes = {
	error: PropTypes.shape({
		showError: PropTypes.bool.isRequired,
		title: PropTypes.string,
		status: PropTypes.number,
		message: PropTypes.string,
	}),
	handleClearError: PropTypes.func.isRequired
}

export default ErrorMessage;
