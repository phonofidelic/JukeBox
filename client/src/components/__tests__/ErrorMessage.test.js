import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { createRender } from '@material-ui/core/test-utils';
import ErrorMessage from '../ErrorMessage';

const INITIAL_STATE = {
	showError: false,
	data: null,
	title: null,
	status: null,
	message: null,
};

describe('ErrorMessage', () => {
	let props;
	let mountedErrorMessage;
	let muiRenderer;

	const errorMessage = props => {
		if (!mountedErrorMessage) {
			mountedErrorMessage = mount(
				<ErrorMessage {...props} />
			);
		}
		return mountedErrorMessage;
	};

	beforeEach(() => {
		props = {
			error: INITIAL_STATE,
			handleClearError: jest.fn()
		};
		mountedErrorMessage = undefined;
		muiRenderer = createRender();
	});

	it('renders an empty div if no error is present', () => {
		const tree = renderer
			.create(<ErrorMessage {...props} />)
			.toJSON();
		const divs = errorMessage(props).find('div');

		expect(tree).toMatchSnapshot();
		expect(divs).toHaveLength(1);
	});

	/***
	 *	Snapshot testing material-ui components needs createRender
	 *	for tesing rendered Dialog component:
	 *	https://material-ui.com/guides/testing/#createrender-options-render
	 *
	 *	This is needed because material-ui uses React portals to
	 *	render modals: 
	 *	https://reactjs.org/docs/portals.html
	 *	https://github.com/facebook/react/issues/11565
	 */
	it('renders a Dialog if an error is present', () => {
		props.error = {
			showError: true,
			title: 'Test Error',
			status: 418,
			message: 'This is a test.'
		};

		const tree = muiRenderer(<ErrorMessage {...props} />);
		const dialog = errorMessage(props).find('Dialog');
		const dialogTitle = errorMessage(props).find('DialogTitle');
		const dialogContent = errorMessage(props).find('DialogContent');
		const dialogActions = errorMessage(props).find('DialogActions');

		expect(tree).toMatchSnapshot();
		expect(dialog).toHaveLength(1);
		expect(dialogTitle).toHaveLength(1);
		expect(dialogContent).toHaveLength(1);
		expect(dialogActions).toHaveLength(1);
	});
});

