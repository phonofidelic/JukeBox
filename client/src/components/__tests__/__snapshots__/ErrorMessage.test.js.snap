// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage renders a Dialog if an error is present 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`ErrorMessage renders an empty div if no error is present 1`] = `<div />`;
